### PR TITLE
Add `from_reader` to `DeserializableSlice`

### LINKED ISSUE #11: From slice take and split
A recurrent pattern is when we have multiple serial structs in a single slice of bytes.

One efficient solution is to have a function `fn from_slice_split<T>(bytes: &[u8]) -> Result<(T, &[u8]), _>`

It will take the first `n` bytes required to parse the provided type, split the slice and return a tuple with the parsed type and the remainder bytes.

Considering the required traits are implemented for T and we have a tuple `(T, T) = (a, b)`, the following test case should pass:
```rust
const B: u8 = 0xfa;

let mut bytes = [B; 2 * T::SIZE + 1];
bytes[..T::SIZE].copy_from_slice(&a.to_bytes());
bytes[T::SIZE..2 * T::SIZE].copy_from_slice(&b.to_bytes());

let (a_p, bytes) = T::from_slice_split(&bytes).unwrap();
let (b_p, bytes) = T::from_slice_split(&bytes).unwrap();

assert_eq!(a, a_p);
assert_eq!(b, b_p);
assert_eq!(bytes, [B]);
```

### PR DESCRIPTION
**Pull request by ZER0**
- Add `serialize::Read` trait

Resolves: #11

### CODE REVIEW

File: dusk-bytes/src/serialize.rs [hunk 1]
```diff
@@ -42,8 +42,56 @@ pub trait DeserializableSlice<const N: usize>: Serializable<N> {
...
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        if buf.len() > self.len() {
+            return Err(Error::bad_length(self.len(), buf.len()));
+        }
+        let amt = buf.len();
+        let (a, b) = self.split_at(amt);
+
+        // First check if the amount of bytes we want to read is small:
+        // `copy_from_slice` will generally expand to a call to `memcpy`, and
+        // for a single byte the overhead is significant.
```
Comment (CPerezz) [inline dusk-bytes/src/serialize.rs:87]: More significant than adding branching to the CPU?

**Review by CPerezz — APPROVED** [review]
LGTM

File: dusk-bytes/src/serialize.rs (hunk 1 above)
Comment (ZER0) [reply dusk-bytes/src/serialize.rs:87]: Definitely yes.
And this is come from the Rust language source code btw. :)

File: dusk-bytes/src/serialize.rs [hunk 2]
```diff
@@ -42,8 +42,56 @@ pub trait DeserializableSlice<const N: usize>: Serializable<N> {
...
+        if amt == 1 {
```
Comment (vlopes11) [inline dusk-bytes/src/serialize.rs:88]: Its fine here, but in our use-case we will (almost) never fall into serialization / deserialization of single bytes. Hence, we will, in practice, never benefit from this part of the code.

But, as a general purpose crate, it makes sense.

**Review by vlopes11 — APPROVED** [review]
LGTM, this will help us a lot

File: dusk-bytes/src/serialize.rs (hunk 2 above)
Comment (ZER0) [reply dusk-bytes/src/serialize.rs:88]: We have at least the `NoteType` that is serialized as single byte, and I think we could have similar usage for indicates type / flags as well in the future.

In this way we don't have to make assumption (as we do now), all the types can use `from_reader` (even `NoteType`) - instead that deserialize manually, and the code will be more consistent.

**Pull request merged**